chore: Clarify error boundaries in instrumentation spec (#12822)
vectordotdev/vector: docs/specs/component.md

<!-- MarkdownTOC autolink="true" style="ordered" indent="   " -->

- [Introduction](#introduction)
- [Scope](#scope)
- [How to read this document](#how-to-read-this-document)
- [Naming](#naming)
  - [Source and sink naming](#source-and-sink-naming)
  - [Transform naming](#transform-naming)
- [Configuration](#configuration)
  - [Options](#options)

>     - [`endpoint(s)`](#endpoints)

Reviewer: It wasn't changed here, but since it caught my eye: do we really want this to be a MUST in a component's configuration? Like, would we consider this for something like.... the S3 sink, or GCS sink, where the endpoint is well-known based on a combination of the region, bucket name, etc?

Author: Anyway, I agree, but I don't want to this to block getting this merged, so resolving for now.

#### `endpoint(s)`

When a component makes a connection to a downstream target, it SHOULD
expose either an `endpoint` option that takes a `string` representing a
single endpoint, or an `endpoints` option that takes an array of strings

> representing multiple endpoints. If a component uses multuiple options to

Reviewer 2: ```suggestion
representing multiple endpoints. If a component uses multiple options to
```

## Instrumentation

**Extends the [Instrumentation Specification].**

Vector components MUST be instrumented for optimal observability and monitoring.

### Events

> This section lists all required and optional events that a component MUST emit.

Reviewer: ```suggestion
This section lists all required events that a component MUST emit. Additionally, additional events are listed that a component is RECOMMENDED to emit, but remain OPTIONAL.
```

> This section lists all required events that a component MUST emit. Additionally, additional events are listed that a component is RECOMMENDED to emit, but remain OPTIONAL.

Reviewer: Sorry, I can't believe I didn't notice that this was in the suggestion I wrote... 🤦🏻 

```suggestion
This section lists all required events that a component MUST emit. Additional events are
```

It is expected that components will emit custom events beyond those listed here
that reflect component specific behavior. There is leeway in the implementation

> of these events:

Reviewer: ```suggestion
This section lists all required events that a component MUST emit. Additional events
are listed that a component is RECOMMENDED to emit, but remain OPTIONAL. It is
expected that components will emit custom events beyond those listed here that
reflect component specific behavior. There is leeway in the implementation of these
events:

Reviewer: Sorry, this was another casualty of one of my previous suggestions getting cut off.

  individual events. For example, emitting the `EventsReceived` event for 10
  events MUST increment the `component_received_events_total` counter by 10.

#### ComponentBytesReceived

> *Sources* MUST emit a `BytesReceived` event immediately after receiving, decompressing

Reviewer 2: ```suggestion
*Sources* MUST emit a `ComponentBytesReceived` event immediately after receiving, decompressing
```

    the defined properties as metric tags.
* Logs
  * MUST log a `Bytes received.` message at the `trace` level with the
    defined properties as key-value pairs.
  * MUST NOT be rate limited.

#### ComponentEventsReceived

> *All components* MUST emit an `EventsReceived` event immediately after creating

Reviewer 2: ```suggestion
*All components* MUST emit an `ComponentEventsReceived` event immediately after creating
```

    `byte_size` property with the other properties as metric tags.
* Logs
  * MUST log a `Events received.` message at the `trace` level with the
    defined properties as key-value pairs.
  * MUST NOT be rate limited.

#### ComponentEventsSent

*All components* that send events down stream, and delete them in Vector, MUST

> emit an `EventsSent` event immediately after sending, if the transmission was

Reviewer 2: ```suggestion
emit an `ComponentEventsSent` event immediately after sending, if the transmission was
```

    in JSON representation.
* Logs
  * MUST log a `Events sent.` message at the `trace` level with the
    defined properties as key-value pairs.
  * MUST NOT be rate limited.

#### ComponentBytesSent

*Sinks* that send events down stream, and delete them in Vector, MUST emit

> a `BytesSent` event immediately after sending bytes to the downstream target, if

Reviewer 2: ```suggestion
a `ComponentBytesSent` event immediately after sending bytes to the downstream target, if
```

    defined properties as metric tags.
* Logs
  * MUST log a `Bytes sent.` message at the `trace` level with the
    defined properties as key-value pairs.
  * MUST NOT be rate limited.

#### ComponentError

**Extends the [Error event].**

*All components* MUST emit error events in accordance with the [Error event]

> requirements.

Author: Shifted the `*Error` event requirements to the instrumentation spec since this is shared across Vector.

This specification does not list a standard set of errors that components must
implement since errors are specific to the component.

#### ComponentEventsDropped

**Extends the [EventsDropped event].**

*All components* that can drop events MUST emit a `ComponentEventsDropped`

> event in accordance with the [EventsDropped event] requirements.

Author: Shifted the `*EventsDropped` event requirements to the instrumentation spec since this is shared across Vector.

    of Vector events discarded if the error resulted in discarding (dropping)
    acknowledged events.
    * For sources, only increment this metric if incoming events were consumed
      (and acknowledged if applicable) and discarded. The metric MUST not
      include events that will be re-ingested.
    * For sinks, this means only incrementing this metric if the error resulted
      in the sink dropping the events, and thus acknowledging them. Retried
      events MUST not be included in the metric.
* Logs
  * If `recoverable` is `false`, MUST log a message at the `error` level.
    Otherwise, if `recoverable` is `true`, MUST log a message at the `warn`
    level.
  * MUST include the defined properties as key-value pairs.
  * SHOULD be rate limited to 10 seconds.

#### ComponentEventsDropped

*All components* that can drop events, **unrelated to an error**, MUST emit a
`ComponentEventsDropped` event. Otherwise, if an event is dropped due to an
error, then the error event itself MUST emit the appropriate logs and metrics
as described in the [ComponentError](#ComponentError) section.

* Metrics
  * MUST increment the `component_discarded_events_total` counter by the number

>     of Vector events discarded.

Reviewer 2: How will users differentiate between events dropped intentionally (like by the `filter` transform) and those dropped due to an error? Also, what tags should be applied? Should it have the same `stage` tag as the error `component_discarded_events_total` has?

Author: I think it's the presence of the `error_type` tag, right? I still think it's somewhat strange that the `component_discarded_events_total` has an `error_type` tag, but that's how it works now. So I think it's fine in the short term.

Reviewer 2: 🤔 that feels to me like it is overloading the `error_type` tag a bit. Maybe we could add a `reason` tag that is `error|intentional`?

Author: Hm, can you think of a case where we'd unintentionally drop data not due to an error?

The only one I can think of is an event that never matches any of the routes in the `route` transform, but that could be an error.

Reviewer 2: Yeah, I was thinking transforms like `route`, `sample`, and `throttle`.

Author: Resolved and moved the discussion [here](https://github.com/vectordotdev/vector/pull/12822#discussion_r880782837).
